internal/eventstream/client.go

```diff
@@ -0,0 +1,415 @@
+		} else {
+			client.Logger.Infow("Replaying the API has finished", zap.Duration("duration", time.Since(startTime)))
+		}
+
+		close(finCh)
+	}()
+
+	return finCh
+}
+
+// worker is the Client's main background worker, taking care of event.Event dispatching and mode switching.
+//
+// When the Client is in the replay phase, requested by replayPhaseRequest, events from the Event Stream API will
+// be cached until the replay phase has finished, while replayed events will be delivered directly.
+//
+// Communication takes place over the channels eventDispatcherEventStream, eventDispatcherReplay and replayPhaseRequest.
+func (client *Client) worker() {
+	// eventCache is a subset of event.Event used for caching those in replayCache below.
+	type eventCache struct {
+		SourceId int64
+		Name     string
+		Type     string
+	}
+
+	var (
+		// replayFinCh holds a reference when the Client is in the replay phase. It will be closed from the producer,
+		// startReplayWorker, when replaying is finished, which indicates the select below to switch phases. When this
+		// variable is nil, the Client is in the normal operating phase.
+		replayFinCh chan struct{}
+
+		// replayBuffer holds Event Stream events to be replayed after the replay phase has finished.
+		replayBuffer = make([]*event.Event, 0)
+		// replayCache maps eventCache derived from event.Events to API time to skip replaying outdated events.
+		replayCache = make(map[eventCache]time.Time)
+	)
+
+	// replayCacheUpdate updates the replayCache if this eventMsg seems to be the latest of its kind.
+	replayCacheUpdate := func(ev *eventMsg) {
+		evc := eventCache{ev.event.SourceId, ev.event.Name, ev.event.Type}
+		ts, ok := replayCache[evc]
+		if !ok || ev.apiTime.After(ts) {
+			replayCache[evc] = ev.apiTime
+		}
+	}
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing down main worker as context is finished", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.replayPhaseRequest:
+			if replayFinCh != nil {
+				// There shouldn't be multiple concurrent startReplayWorker calls. However, technically this is possible,
+				// i.e. when the Icinga 2 API connection is flapping.
+				client.Logger.Error("Another replay phase request was sent while the Client is already replaying")
+				break
+			}
+
+			client.Logger.Debug("Dispatcher enters replay phase, starting caching Event Stream events")
+			replayFinCh = client.startReplayWorker()
+
+		case <-replayFinCh:
+			skipCounter := 0
+			for _, ev := range replayBuffer {
+				ts, ok := replayCache[eventCache{ev.SourceId, ev.Name, ev.Type}]
+				if ok && ev.Time.Before(ts) {
+					client.Logger.Debugw("Skip replaying outdated Event Stream event", zap.Stringer("event", ev),
+						zap.Time("event timestamp", ev.Time), zap.Time("cache timestamp", ts))
+					skipCounter++
+					continue
+				}
+
+				client.CallbackFn(ev)
+			}
```

Comment:
While this loop is running (which may take a bit given that it performs database operations), no more receives from `client.eventDispatcherEventStream` take place, which could block further reads from the event stream response. So maybe move this into a goroutine or, as that would probably get annoying given that you'd then have to synchronize access on `replayCache`, restructure the case so that it processes the first event, and loops until the buffer is empty.

Reply:
Valid point. To not have to worry about huge callback delays when working on multiple cached events, I am going to implement a second independent FIFO-like queue which gets feed with events where currently `CallbackFn` is called.